### PR TITLE
Db

### LINKED ISSUE #7: Add DB
For storing last downloaded version and creating delta packages

### PR DESCRIPTION
**Pull request by nirolevy**
Closes #7

### CODE REVIEW

**Comment by Guinness1759** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 16
           

Complexity increasing per file
==============================
- src/DAL/MongoDB.py  3
- src/import_util.py  6
- src/DAL/_DAL.py  1
- src/DAL/FS.py  5
- src/DAL/__init__.py  1
- src/providers/_Provider.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/mh.guinness/vsix-downloader/pullRequest?prid=5655161&bid=18514333)

File: __main__.py [hunk 1]
```diff
@@ -3,47 +3,74 @@
...
+                logger.info(f'Started resolving products from provider {provider_name}')
+                provider = providers[provider_name]
+                requested_products = [ (product.name, product.version) for product in db.get_products(provider) ]
+                logger.info(f'Found {len(requested_products)} products in DB for provider {provider_name}')
+                provider_products = provider(logger).provide(requested_products)
+                resolved_products += provider_products
+                logger.info(f'Resolved {len(provider_products)} products for provider {provider_name}')
+        logger.info('Successfully resolved products')
+
+        resolved_products = [ product for product in resolved_products if not db.is_downloaded_before(product) ]
```
Comment (syncush) [inline __main__.py:57]: if we set up mongo, we would query mongo for each `product`, assuming we use a managed `mongodb` the round trip is costly.

File: __main__.py [hunk 2]
```diff
@@ -3,47 +3,74 @@
...
+            return
+
+        logger.info('Started packaging resolved products')
+        packager = packagers.get_packager(args.packager, **args_dict, logger=logger, num_workers=num_workers)
+        packager.package(resolved_products)
+        logger.info('Successfully packaged products')
+
+        logger.info('Reporting to DB')
+        for product in resolved_products:
+            db.add_resolved_product(product)
```
Comment (syncush) [inline __main__.py]: can we do it in a bulk?

File: src/command_line.py [hunk 3]
```diff
@@ -20,21 +21,31 @@
...
-    for code, packager_type in packager_type_codes.items()
-]
-__parser.add_argument('--packager', '-p', type=str, default=default_packager_type_code, choices=packager_type_codes.keys(),
-                      help='Which packager to use. options are:\n{0}'.format('\n'.join(packager_type_helps)))
+# Database
+__parser.add_argument('--database', type=str, default=DAL.default,
+                      choices=[db.name for db in DAL.get_databases()],
+                      help='Specify used database')
+__parser.add_argument('--database-args', type=str,
+                      help='Specify arguments for used database')
```
Comment (syncush) [inline src/command_line.py:33]: can you supply an example of how do we set it over MongoDB?

File: src/command_line.py [hunk 4]
```diff
@@ -4,8 +4,9 @@
 from str2bool import str2bool
 
 from src import __version__ as version
+from src import DAL
 from src.packagers import packager_type_codes, default_packager_type_code
-from src.providers import providers
+from src.providers import get_providers
 
 
 __parser = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter)
```
Comment (syncush) [inline src/command_line.py:12]: we need to upsert README.md

File: __main__.py (hunk 1 above)
Comment (nirolevy) [reply __main__.py:57]: If I understand you right, you want to filter when receiving the initial list of products to download. But this happens after the providers' resolution... If you'd like, I can parallelize it

Comment (nirolevy) [reply __main__.py:57]: Besides, round trip shouldn't be THAT costly, it's just sending and receiving a single product dictionary

File: src/command_line.py (hunk 3 above)
Comment (nirolevy) [reply src/command_line.py:33]: It's the connection string (i.e. `mongodb+srv://<username>:<password>@<instance>/<db>`). When We'll write the configuration loader (which loads `.yaml` files, `.json` files, `.env` files, env variables and CLI parameters) we'll incorporate into that examples and better explenations

File: src/command_line.py (hunk 4 above)
Comment (nirolevy) [reply src/command_line.py:12]: Huh?

File: __main__.py (hunk 2 above)
Comment (nirolevy) [reply __main__.py]: Yes

**Comment by Guinness1759** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 15
           

Complexity increasing per file
==============================
- src/DAL/FS.py  5
- src/import_util.py  6
- src/providers/_Provider.py  4
- src/DAL/MongoDB.py  3
- src/DAL/_DAL.py  1
- src/DAL/__init__.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/mh.guinness/vsix-downloader/pullRequest?prid=5655161&bid=18514333)

File: __main__.py (hunk 1 above)
Comment (syncush) [reply __main__.py:57]: Scaling up for 1548 npm packages alone require many http calls.

I would like to resolve all the products in a. Single mongodb pipeline aggregation

File: src/command_line.py (hunk 4 above)
Comment (syncush) [reply src/command_line.py:12]: Github code line marking is a bit buggy, what i meant is that we need to update our readme, you have introduced a whole new cli arguments. We must keep track with how one can use that tool

File: __main__.py (hunk 1 above)
Comment (nirolevy) [reply __main__.py:57]: That's data from the server that you want to check if it exists in the DB seperately per document. How would you filter over it and where exactly will aggregations come into play?

File: src/command_line.py (hunk 4 above)
Comment (nirolevy) [reply src/command_line.py:12]: When We'll write the configuration loader (which loads .yaml files, .json files, .env files, env variables and CLI parameters) we'll incorporate into that examples and better explenations

Comment (nirolevy) [reply src/command_line.py:12]: Maybe even generate a '.md' file

File: __main__.py (hunk 1 above)
Comment (nirolevy) [reply __main__.py:57]: Keep in mind that it's a single TCP connection being reused and almost no unnecessary data is passed (can be reduced to absolutely no unnecessary data if I use 'count_documents')

File: src/command_line.py (hunk 4 above)
Comment (syncush) [reply src/command_line.py:12]: Even a small, simple README is fine for now

File: __main__.py (hunk 1 above)
Comment (syncush) [reply __main__.py:57]: Alright, I agree, truly almost no unnecessary data is passed, resolved

File: src/command_line.py (hunk 4 above)
Comment (nirolevy) [reply src/command_line.py:12]: I think that is out of scope and makes the `readme` dependant of the code itself, but as you wish

**Comment by Guinness1759** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 16
           

Complexity increasing per file
==============================
- src/DAL/MongoDB.py  3
- src/DAL/FS.py  5
- src/DAL/__init__.py  1
- src/providers/_Provider.py  4
- src/DAL/_DAL.py  1
- src/import_util.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/mh.guinness/vsix-downloader/pullRequest?prid=5655161&bid=18514333)

**Review by syncush — APPROVED** [review]

**Pull request merged**